jraft-core/src/main/java/com/alipay/sofa/jraft/util/SegmentList.java

```diff
@@ -0,0 +1,347 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.alipay.sofa.jraft.util;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.function.Predicate;
+
+/**
+ * A list implementation based on segments.Only supports removing elements from start or end.
+ * The list keep the elements in a segment list, every segment contains at most 128 elements.
+ *
+ * @author boyan(boyan@antfin.com)
+ *
+ */
+public class SegmentList<T> {
+    private static final int             SEGMENT_SHIFT = 8;
+    public static final int              SEGMENT_SIZE  = 2 << (SEGMENT_SHIFT - 1);
```

Comment:
every segment contains at most 128 elements, but **SEGMENT_SIZE 2 << 7 equals 256**.

Reply:
Fixed.  128 is a balance between time and space.